modules/decrypt-node/test/decipher_stream.test.ts

```diff
@@ -0,0 +1,132 @@
+  })
+
+  it('Precondition: Ciphertext must not be flowing before a BodyHeader is processed.', () => {
+    const verifyStream = new VerifyStream({})
+    const test = getDecipherStream()
+    verifyStream.pipe(test)
+    verifyStream.emit('DecipherInfo', {
+      messageId: Buffer.from(Array(16)),
+      contentType: ContentType.FRAMED_DATA,
+      iv: Buffer.from(Array(12)),
+      getDecipher: () => ({
+        setAAD () {
+          // set decipher to a true-ish value for test
+          return true
+        }
+      })
+    })
+    const bodyInfo = {
+      contentLength: 123,
+      iv: Buffer.from(Array(12)),
+      sequenceNumber: 1,
+      isFinalFrame: false
+    }
+    test._onBodyHeader(bodyInfo)
+
+    // The basis of this test is that 2 bodyInfo events must not
+    // happen without a AuthTag in between
+    expect(() => test._onBodyHeader(bodyInfo)).to.throw('Malformed State.')
+  })
+
+  it('Precondition: BodyHeader must be parsed before frame data.', () => {
+    const test = getDecipherStream()
+    expect(() => test._transform(Buffer.from([1]), 'binary', () => {})).to.throw('Malformed State.')
+  })
+
+  it('Precondition: Only content should be transformed, so the lengths must always match.', () => {
+    const verifyStream = new VerifyStream({})
+    const test = getDecipherStream()
+    verifyStream.pipe(test)
+    verifyStream.emit('DecipherInfo', {
+      messageId: Buffer.from(Array(16)),
+      contentType: ContentType.FRAMED_DATA,
+      iv: Buffer.from(Array(12)),
+      getDecipher: () => ({
+        setAAD () {
+          // set decipher to a true-ish value for test
+          return true
+        }
+      })
+    })
+    const bodyInfo = {
+      contentLength: 0,
+      iv: Buffer.from(Array(12)),
+      sequenceNumber: 1,
+      isFinalFrame: false
+    }
+    test._onBodyHeader(bodyInfo)
+
+    expect(() => test._transform(Buffer.from([1]), 'binary', () => {})).to.throw('Lengths do not match')
+  })
+
+  it('Precondition: _onAuthTag must be called only after a frame has been accumulated.', async () => {
+    const verifyStream = new VerifyStream({})
+    const test = getDecipherStream()
+    verifyStream.pipe(test)
+    verifyStream.emit('DecipherInfo', {
+      messageId: Buffer.from(Array(16)),
+      contentType: ContentType.FRAMED_DATA,
+      iv: Buffer.from(Array(12)),
+      getDecipher: () => ({
+        setAAD () {
+          // set decipher to a true-ish value for test
+          return true
+        }
+      })
+    })
+    const bodyInfo = {
+      contentLength: 10,
+      iv: Buffer.from(Array(12)),
+      sequenceNumber: 1,
+      isFinalFrame: false
+    }
+    test._onBodyHeader(bodyInfo)
+
+    await expect(test._onAuthTag(Buffer.from([]), () => {})).to.rejectedWith(Error, 'AuthTag before frame.')
+  })
+
+  it('Precondition: I must have received all content for this frame.', async () => {
+    /* The fact that I can not figure out how to test this,
```

Comment:
Let's cut an issue tracking figuring this out and link it here?

Reply:
It is not really testable.  I'm noting this in code and handling this differently.
Basically there is a closure that holds this state.  And I do not see _how_, short of magically changing raw RAM, to make this fail in order to test it...